### PR TITLE
[WIP] Split xcconfig to pod_target_xcconfig and user_target_xcconfig

### PR DESCRIPTION
**Pull request by mrackwitz**
* [x] Implement the deprecation logic.

Belongs to CocoaPods/CocoaPods#3573.

### CODE REVIEW

**Comment by segiddins** [comment]
Can we think of a more concise name for those attributes? Think along the lines of public/private.

**Comment by segiddins** [comment]
What's the reason for keeping the user target xcconfig option at all?

File: lib/cocoapods-core/specification/dsl/deprecations.rb [hunk 1]
```diff
@@ -9,6 +9,13 @@ def preferred_dependency=(name)
           CoreUI.warn "[#{self}] `preferred_dependency` has been renamed "\
             'to `default_subspecs`.'
         end
+
+        def xcconfig=(name)
```
Comment (segiddins) [inline lib/cocoapods-core/specification/dsl/deprecations.rb]: For backwards compatibility, won't this need to set both?

**Comment by mrackwitz** [comment]
> Can we think of a more concise name for those attributes? Think along the lines of public/private. 

I was thinking about that, but in my opinion, this is just masking the meaning as you would usually expect that something public is also visible/accessible from the private area. But those here shouldn't be inherited like it happens for `PublicPodXCConfig`. Those names used here are very explicit about the effect range.

>What's the reason for keeping the user target xcconfig option at all?

I think this discussion should go in CocoaPods/CocoaPods#3465.

File: lib/cocoapods-core/specification/dsl/deprecations.rb (hunk 1 above)
Comment (mrackwitz) [inline lib/cocoapods-core/specification/dsl/deprecations.rb]: Good point! :+1:

**Comment by mrackwitz** [comment]
@orta: Can you review [the documentation of the new attribute](https://github.com/CocoaPods/Core/pull/242/files#diff-0c2cb09f968f1dd4c986d5e00ac9a4aeR737)?

File: lib/cocoapods-core/specification/dsl/deprecations.rb [hunk 2]
```diff
@@ -9,6 +9,13 @@ def preferred_dependency=(name)
           CoreUI.warn "[#{self}] `preferred_dependency` has been renamed "\
             'to `default_subspecs`.'
         end
+
+        def xcconfig=(value)
+          self.pod_target_xcconfig = value
+          self.user_target_xcconfig = value
+          CoreUI.warn "[#{self}] `xcconfig` has been splitted to "\
```
Comment (segiddins) [inline lib/cocoapods-core/specification/dsl/deprecations.rb]: `has been split between`

File: lib/cocoapods-core/specification/dsl.rb [hunk 3]
```diff
@@ -717,18 +717,53 @@ def dependency(*args)
...
+      #           Key-value pairs representing build settings.
       #
-      attribute :xcconfig,
+      attribute :pod_target_xcconfig,
+                :container => Hash,
+                :inherited => true
+
+      # @!method user_target_xcconfig=(value)
+      #
+      #   Any flag to add to the final aggregate target xcconfig file, which
```
Comment (orta) [inline lib/cocoapods-core/specification/dsl.rb]: > Any flag to add

Specifies flags to add to

File: lib/cocoapods-core/specification/dsl.rb [hunk 4]
```diff
@@ -717,18 +717,53 @@ def dependency(*args)
...
+      #   propagates to non-overridden build settings to the integrated user
+      #   targets.
+      #
+      #   ---
+      #
+      #   This attribute is __not recommended__ as Pods should not pollute the
+      #   build settings of the user project and this can cause conflicts.
+      #
+      #   Multiple definitions for build settings, which take multiple values,
```
Comment (orta) [inline lib/cocoapods-core/specification/dsl.rb]: > for build settings, which take multiple values,

for build settings that take multiple values 

( remove the comma too )

File: lib/cocoapods-core/specification/dsl.rb [hunk 5]
```diff
@@ -717,18 +717,53 @@ def dependency(*args)
...
+      #   will be merged. The user is warned on conflicting definitions for
+      #   custom build settings and build settings, which take only one value.
```
Comment (orta) [inline lib/cocoapods-core/specification/dsl.rb]: > settings, which

settings that

File: lib/cocoapods-core/specification/dsl.rb [hunk 6]
```diff
@@ -717,18 +717,53 @@ def dependency(*args)
...
+      #
+      #   Typically clang compiler flags or precompiler macro definitions go
+      #   in here, if they are required when importing the pod in the user
```
Comment (orta) [inline lib/cocoapods-core/specification/dsl.rb]: > in here, if they

in here if they

File: lib/cocoapods-core/specification/dsl.rb [hunk 7]
```diff
@@ -717,18 +717,53 @@ def dependency(*args)
...
+      #   target. Note that, this influences only the compiler view of the
```
Comment (orta) [inline lib/cocoapods-core/specification/dsl.rb]: > this influences only the compiler view

this influences not only the compiler view

File: lib/cocoapods-core/specification/dsl.rb [hunk 8]
```diff
@@ -717,18 +717,53 @@ def dependency(*args)
...
+      #   public interface of your pod, but also all other integrated pods
+      #   alongside to yours. You should always prefer [`pod_target_xcconfig`](
+      #   http://guides.cocoapods.org/syntax/podspec.html#pod_target_xcconfig),
+      #   which can contain the same settings, but only influence the
+      #   toolchain, when compiling your pod target.
```
Comment (orta) [inline lib/cocoapods-core/specification/dsl.rb]: > toolchain, when compiling

toolchain when compiling

**Comment by orta** [comment]
done'd - thanks for the shout out

**Comment by mrackwitz** [comment]
@orta: Thank you for all the feedback. It's addressed now.

File: lib/cocoapods-core/specification/dsl.rb [hunk 9]
```diff
@@ -717,18 +717,53 @@ def dependency(*args)
...
+      attribute :pod_target_xcconfig,
+                :container => Hash,
+                :inherited => true
+
+      # @!method user_target_xcconfig=(value)
+      #
+      #   Specifies flags to add to the final aggregate target xcconfig file,
+      #   which propagates to non-overridden and inheriting build settings to
+      #   the integrated user targets.
+      #
```
Comment (orta) [inline lib/cocoapods-core/specification/dsl.rb]: > Specifies flags to add to the final aggregate target xcconfig file. These can be overridden by a library consumer. During integration CocoaPods will warn the consumer if the build setting does not contain `$(inherited)` if they have been changed.

**Comment by mrackwitz** [comment]
I just realized that the deprecations only work for the Ruby DSL. JSON podspecs remain unaffected from that, which would mean that we would need to re-write all specs and instantly loose backwards-compatibility or we introduce deprecation logic into the `Specification` model. I tend to prefer the latter. Any thoughts on that? /c @kylef

**Comment by orta** [comment]
+1 on deprecation logic

**Comment by segiddins** [comment]
@mrackwitz :+1: on deprecation logic when reading from the `Specification`. Please also add a lint step that validates that `xcconfig` and the public/private ones are used mutually exclusively?

**Comment by mrackwitz** [comment]
@segiddins: Good idea!

File: lib/cocoapods-core/specification/dsl/deprecations.rb [hunk 10]
```diff
@@ -9,6 +9,13 @@ def preferred_dependency=(name)
           CoreUI.warn "[#{self}] `preferred_dependency` has been renamed "\
             'to `default_subspecs`.'
         end
+
+        def xcconfig=(value)
+          self.pod_target_xcconfig = value
+          self.user_target_xcconfig = value
+          CoreUI.warn "[#{self}] `xcconfig` has been split to "\
```
Comment (segiddins) [inline lib/cocoapods-core/specification/dsl/deprecations.rb]: `into`

**Comment by segiddins** [comment]
@mrackwitz please review my latest commit

**Comment by segiddins** [comment]
@mrackwitz we good to merge this?

**Comment by mrackwitz** [comment]
@segiddins: Let me merge the PR on Core, rebase this, update the bundle and rebuild the integration specs, then we are good to go.

**Comment by segiddins** [comment]
:+1: I'll handle that right now :)

**Comment by kylef** [comment]
@segiddins are you still handling this?

https://github.com/CocoaPods/CocoaPods/pull/3573 hasn't been merged yet.

**Comment by segiddins** [comment]
@kylef no, @mrackwitz told me he wanted to handle this

**Comment by mrackwitz** [comment]
I did. https://github.com/CocoaPods/CocoaPods/pull/3573 should be ready now.
But I didn't realized earlier that we need to get that merged before anything else to maintain a releasable `master` on the CocoaPods repo.

**Pull request merged**